### PR TITLE
ONNX importer implementation

### PR DESCRIPTION
**Pull request by aclex**
As discussed in #87 I've been preparing the implementation of ONNX importer for nncase. It's pretty much based on the other existing importers, mainly `tflite` and `caffe` ones, but I've also tried to add support for those ONNX operators, which have direct counterparts in the nncase IR. To be sure it works properly a test case suite is also added, similar to the one for `tflite` part. It is run by `pytest` out of the root directory. It relies on PyTorch as a source code for ONNX snippet models. 

Here's a table of the operators implemented and tested:
|ONNX operator | Covered by tests|
|-------------------|----------|
|Abs|✅|
|Add|✅|
|AveragePool|—|
|Ceil|✅|
|Concat|✅|
|Constant|—|
|Conv|✅|
|ConvTranspose|✅|
|Cos|✅|
|Div|✅|
|Exp|✅|
|Floor|✅|
|LpNormalization|—|
|LeakyRelu||
|Log|✅|
|MatMul|✅|
|MaxPool|—|
|Max|✅|
|Min|✅|
|Mul|✅|
|Neg|✅|
|Pad|✅|
|ReduceMax|✅|
|ReduceMean|✅|
|ReduceMin|✅|
|ReduceSum|✅|
|Relu||
|Reshape|✅|
|Resize|✅|
|Sqrt|✅|
|Sin|✅|
|Slice|✅|
|Softmax|—|
|Sub|✅|
|Transpose|✅|

I'm trying to finish the tests implementation (and of course the fixes along it :)) so although the changes are apparently not yet ready to be merged, I hope to finish it all soon. Unfortunately, it's not yet possible to reflect any operator version of ONNX completely to be absolutely safe while exporting and importing your models, but the target version is 11 and the range of ONNX operators used in export can usually be more or less controlled by choosing the certain model structure.

There're also plans to implement `Gemm` and `BatchNormalization` operators in the emulated form due to their popularity noticed in the models out there.

Please feel free to comment and review the code. As I said, my first aim is to finish the tests and fixes, but then I'd be glad to discuss and make all the necessary fixes and corrections.

### CODE REVIEW

**Comment by aclex** [comment]
@sunnycase finally I think I'm done with the test cases, tried to add as much test coverage, as possible, but at times it's not so possible due to PyTorch bugs in ONNX export or some other limitations. Generally I hope it's fine to be merged now, if you would find it useful or possible. I'll still continue to test it with the complete models now, want to run some popular CV ones, but any additional bug fixes can hopefully go to separate pull requests to not make the process infinite :)

**Comment by sunnycase** [comment]
Thanks. I'll check the compile errors.

**Comment by aclex** [comment]
@sunnycase thank you! It seems, Boost.filesystem is absent somehow (included in the importer base code).

File: src/importer/onnx/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	ops/transpose.cpp
+	ops/unary.cpp
+	)
+
+add_library(onnx_importer ${ONNX_IMPORTER_SOURCES} ${ONNX_IMPORTER_OPS_SOURCES})
+target_compile_definitions(onnx_importer PRIVATE NATIVE_IS_BIG_ENDIAN=${BIG_ENDIAN})
+target_include_directories(onnx_importer PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
+get_filename_component(PARENT_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR} DIRECTORY)
+target_include_directories(onnx_importer PUBLIC ${PARENT_SOURCE_DIR}/include)
+target_link_libraries(onnx_importer PUBLIC hlir)
```
Comment (take-cheeze) [inline src/importer/onnx/CMakeLists.txt]: Seems `CONAN_PKG::boost` is missing
https://github.com/kendryte/nncase/blob/83149a3660ed67f925963b3e1365120d0f9a22eb/src/importer/CMakeLists.txt#L44
```suggestion
target_link_libraries(onnx_importer PUBLIC hlir CONAN_PKG::boost)
```

Comment (aclex) [reply src/importer/onnx/CMakeLists.txt]: Thank you! Indeed it's missing, I'll check it a little later and add your fix.

Comment (aclex) [reply src/importer/onnx/CMakeLists.txt]: Works perfectly, have just added the fix.

**Comment by mirecta** [comment]
wow nice a nd what about operators GATHER UNSQUEEZE
i try compile this mocel ssd detector , tis two is not so complicated ops 
http://server.talasek.sk/mobilenet-v1-ssd-mp-0_675.onnx

**Comment by aclex** [comment]
@mirecta thank you! `Unsqueeze` is indeed also considered to be added. Not sure, if `Gather` is convertible, maybe using slicing, but a little more complex. Anyway, in my opinion, the ONNX importer is still yet to be tested on the real models with the current set of supported operators, and only after this it's fine to add more of them. Still you can dig in the code and give it a try, it's really not so difficult, but requires some spare time.

**Comment by sunnycase** [comment]
To solve the test fails of resize and slice ops, I noticed the exported onnx did'nt fold the constant into resize op. I added a onnx optimization call to `utils.save` but it caused some other issues, so I disabled these tests.

**Comment by mirecta** [comment]
@ so i can help can u get me some intro ? can we us some other communication channel ?

**Comment by aclex** [comment]
@sunnycase Thank you very much for merging it! As for the failing tests of slice and resize, they both work for me, so it might be inconsistencies among PyTorch version (mine is 1.3.1). Maybe it would be better to put the pre-compiled onnx's to the repository and test against them, at least for a number of tests?

**Comment by aclex** [comment]
@mirecta you can contact me via e-mail, but generally I think, I couldn't be more informative, than the source code :) Alternatively you can try to replace the missing ops with something supported in the model itself and try to use this result in the import.

**Comment by corleonechensiyu** [comment]
When I tested the onnx model to kmodel, a **terminate called without an active exception** error occurred.The GDB debugging information is as follows`#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff6cc2801 in __GI_abort () at abort.c:79
#2  0x00007ffff76b5957 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007ffff76bbae6 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff76bbb21 in std::terminate() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x0000555555671766 in tcb::contract_violation () at /home/csy/nncase/third_party/xtl/include/xtl/xspan_impl.hpp:84
#6  0x00005555558f1814 in tcb::span<long const, -1l>::operator[] (this=0x7fffffffce10, idx=1)
    at /home/csy/nncase/third_party/xtl/include/xtl/xspan_impl.hpp:470
#7  0x00005555558f95bd in nncase::importer::onnx_importer::convert_pool (this=0x7fffffffd1a0, node=..., reduce_op=nncase::reduce_max, 
    init_value=3.40282347e+38) at /home/csy/nncase/src/importer/onnx/ops/pool.cpp:118
#8  0x00005555558f8f41 in nncase::importer::onnx_importer::convert_op_MaxPool (this=0x7fffffffd1a0, node=...)
    at /home/csy/nncase/src/importer/onnx/ops/pool.cpp:60
#9  0x00005555558d89b4 in nncase::importer::onnx_importer::convert_op (this=0x7fffffffd1a0, node=...)
    at /home/csy/nncase/src/importer/onnx/opcode.def:17
#10 0x00005555558d7f19 in nncase::importer::onnx_importer::import (this=0x7fffffffd1a0)
    at /home/csy/nncase/src/importer/onnx/onnx_importer.cpp:126
#11 0x00005555558dc178 in nncase::importer::import_onnx (model=...) at /home/csy/nncase/src/importer/onnx/onnx_importer.cpp:776
#12 0x000055555566a417 in (anonymous namespace)::import (options=...) at /home/csy/nncase/src/cli/compile.cpp:130
#13 0x000055555566e134 in compile (options=...) at /home/csy/nncase/src/cli/compile.cpp:404
#14 0x0000555555635f53 in main (argc=12, argv=0x7fffffffdec8) at /home/csy/nncase/src/cli/cli.cpp:45`
as if there was a Maxpool conversion error. My Maxpool information is kernel_shape[2,2], pads[0,0,0,0],strides[2,2].

**Comment by aclex** [comment]
@corleonechensiyu will you be able to provide the input ONNX file to reproduce that?

**Comment by aclex** [comment]
@corleonechensiyu as a sidenote, could you please open a separate issue for the crash to not make the current one really eternal book of wisdom?

**Comment by corleonechensiyu** [comment]
> @corleonechensiyu as a sidenote, could you please open a separate issue for the crash to not make the current one really eternal book of wisdom?
Sorry,my onnx file is [here](https://drive.google.com/file/d/1hCgEvJ0cIqlBUp4rWP46DwhoYXOAFM1N/view?usp=sharing) .I'll delete this issue after u download it.

**Comment by corleonechensiyu** [comment]
Sorry,my onnx file is [here](https://drive.google.com/file/d/1hCgEvJ0cIqlBUp4rWP46DwhoYXOAFM1N/view?usp=sharing) .I'll delete this issue after u download it.

**Comment by aclex** [comment]
@corleonechensiyu got it, thank you! Will try to reproduce it on my package.

**Comment by corleonechensiyu** [comment]
@aclex thank u for your work!

**Comment by aclex** [comment]
@corleonechensiyu sorry for the delay, I've not forgotten about your case, but suffering strange problems with building nncase now, some `absl` dependency seems to not want to link. So I hope to fix it somehow and try to reproduce and possibly debug the problem with your model.

**Comment by corleonechensiyu** [comment]
@aclex thank you

**Comment by aclex** [comment]
@corleonechensiyu to keep you updated on what's going on: I've managed to reproduce the issue with your model and seem to find the cause of it, but it's related to data extraction from the model protobuf and apparently needs some deeper fix, which affects many other parts and elements. I plan to work at it, but it's not going to be a quick fix, as I first expected. If you're still interested, I'll post here as it's finally fixed.

**Comment by corleonechensiyu** [comment]
@aclex thank you.I'm look forward to it

**Comment by krishnak** [comment]
Good work @aclex looks like nncase is growing in its features

**Comment by aclex** [comment]
@krishnak thank you! It's still in quite experimental state now, but I hope it become stable step by step soon.

**Comment by aclex** [comment]
@corleonechensiyu Finally I seem to have managed to finish the fixes in the ONNX importer, submitted them in #123. You can now check the `onnx_ub_fixes` branch in my fork while it's not merged to the trunk: https://github.com/aclex/nncase/tree/onnx_ub_fixes With these changes the model you've sent me goes fine upon the `Gemm` operation, which unfortunately is not yet implemented, but is in my list. Don't know if it would help your experiments, but now it at least don't tend to crash)

**Comment by corleonechensiyu** [comment]
@aclex thank you very much!
